core/server/src/main/java/alluxio/master/file/FileSystemMaster.java

```diff
@@ -2075,63 +2077,97 @@ private long journalPersistedInodes(List<Inode<?>> persistedInodes) {
       mPermissionChecker.checkPermission(Mode.Bits.READ, inodePath);
-      return freeInternal(inodePath, recursive);
+      flushCounter = freeAndJournal(inodePath, options);
+    } finally {
+      // finally runs after resources are closed (unlocked).
+      waitForJournalFlush(flushCounter);
     }
   }
 
   /**
    * Implements free operation.
+   * <p>
+   * This may write to the journal as free operation may change the pinned state of inodes.
    *
    * @param inodePath inode of the path to free
-   * @param recursive if true, and the file is a directory, all descendants will be freed
-   * @return true if the file was freed
+   * @param options options to free
+   * @return the flush counter for journaling
+   * @throws FileDoesNotExistException if the file does not exist
+   * @throws AccessControlException if permission checking fails
+   * @throws InvalidPathException if the given path is invalid
+   * @throws UnexpectedAlluxioException if the file or directory can not be freed
    */
-  private boolean freeInternal(LockedInodePath inodePath, boolean recursive)
-      throws FileDoesNotExistException {
+  private long freeAndJournal(LockedInodePath inodePath, FreeOptions options)
+      throws FileDoesNotExistException, UnexpectedAlluxioException, AccessControlException,
+      InvalidPathException {
     Inode<?> inode = inodePath.getInode();
-    if (inode.isDirectory() && !recursive && ((InodeDirectory) inode).getNumberOfChildren() > 0) {
-      // inode is nonempty, and we don't want to free a nonempty directory unless recursive is
-      // true
-      return false;
+    if (inode.isDirectory() && !options.isRecursive()
+        && ((InodeDirectory) inode).getNumberOfChildren() > 0) {
+      // inode is nonempty, and we don't free a nonempty directory unless recursive is true
+      throw new UnexpectedAlluxioException(ExceptionMessage.CANNOT_FREE_NON_EMPTY_DIR
+          .getMessage(mInodeTree.getPath(inode)));
     }
-
+    long flushCounter = AsyncJournalWriter.INVALID_FLUSH_COUNTER;
+    long opTimeMs = System.currentTimeMillis();
     List<Inode<?>> freeInodes = new ArrayList<>();
     freeInodes.add(inode);
 
-    try (InodeLockList lockList = mInodeTree.lockDescendants(inodePath, InodeTree.LockMode.READ)) {
+    try (InodeLockList lockList = mInodeTree.lockDescendants(inodePath, InodeTree.LockMode.WRITE)) {
       freeInodes.addAll(lockList.getInodes());
-
+      TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath);
       // We go through each inode.
       for (int i = freeInodes.size() - 1; i >= 0; i--) {
         Inode<?> freeInode = freeInodes.get(i);
 
         if (freeInode.isFile()) {
+          if (freeInode.getPersistenceState() != PersistenceState.PERSISTED) {
+            throw new UnexpectedAlluxioException(ExceptionMessage.CANNOT_FREE_NON_PERSISTED_FILE
+                .getMessage(mInodeTree.getPath(freeInode)));
```

Comment:
I guess Calvin's point is that traditionally, the semantics of "force" is to try to complete the operation no matter what.